### DISCUSSION TITLE
PipelineApplicationError: Exception raised in stage [ 0] PdPipelineStage: One-hot encode

### DISCUSSION THREAD

**Discussion by germanlinarespardo** [pdpipe/pdpipe#101] [Q&A] [answered]
- I am trying to set up a pipeline with 5 different steps I want to go through. The steps are the following:

1. One Hot Encoding (ohe)
2. Encoding (other factor variables)
3. Scaling (standard scaler)
4. Lambda function that multiplies my values by some specific weights 
5. Principal Component Analysis (pca)

- For this reason I created a pipeline with these steps in mind within the next function. The pipeline is split accros 2 stages in order to fit the PCA in the first one and apply it in the second one:

```
def proc_file(_df, save_proc=False):
    # Ajustar las columnas
    _df, cols_fact, cols_ohe, dict_weights,cols_num, cols_model = fix_df(_df)
    print(_df.dtypes)
   
    # Identificacion de outliers y limpieza. Extraido de la tuberia para poder implementar la tuberia con los nuevos datos
    if(config.use_DropOutliers):
        _df['outlier'] = P.flag_outliers_row(_df, cols_num, plot=False, save_plot=False, mode=config.OUT_METHOD, pp=config.OUT_IQR_PP, whis=config.OUT_IQR_WHIS, n_sigma=config.OUT_SIGMA_N)
        _df = _df[_df['outlier']==False].drop('outlier',axis=1)
       
    # Es aqui donde hay que generar el modelo PCA, pero implementarlo dentro del PipeLine
    pca = PCA(n_components=config.PCA_nvars)
   
    # Nos montamos nuestro pipeline primero el que hago solamente fit, porque si no en impllment no le va a funcionar
    pipe1 = pdp.PdPipeline([]) # PipeLine de entrenamiento
    if(config.use_OHE):          pipe1 += pdp.OneHotEncode(columns=cols_ohe, exclude_columns=None)    # One Hot Encoding
    if(config.use_FACT):         pipe1 += pdp.Encode(columns=cols_fact, exclude_columns=None)         # Factorize Encoding
    if(config.use_Scaler):       pipe1 += pdp.Scale(config.SCALER)    # Standard Scaler    , exclude_columns = cols_cat + cols_oth)    # Standard Scaler          
    if(config.use_weights):      pipe1 += pdp.AdHocStage(lambda x: apply_weights(x, dict_weights, cols_model, cols_ohe)) # Apply custom weights
    if(config.use_PCA):          pipe1 += pdp.AdHocStage(lambda x: pca.fit(x))           # Se entrena el PCA
   
    pipe1.fit(_df)

    # Nos montamos nuestro pipeline primero el que hago solamente fit, porque si no en impllment no le va a funcionar
    pipe2 = pdp.PdPipeline([]) # Pipeline de ejecucion
    if(config.use_OHE):          pipe2 += pdp.OneHotEncode(columns=cols_ohe, exclude_columns=None)    # One Hot Encoding
    if(config.use_FACT):         pipe2 += pdp.Encode(columns=cols_fact, exclude_columns=None)         # Factorize Encoding
    if(config.use_Scaler):       pipe2 += pdp.Scale(config.SCALER)    # Standard Scaler    , exclude_columns = cols_cat + cols_oth)    # Standard Scaler          
    if(config.use_weights):      pipe2 += pdp.AdHocStage(lambda x: apply_weights(x, dict_weights, cols_model, cols_ohe)) # Apply custom weights
    if(config.use_PCA):          pipe2 += pdp.AdHocStage(lambda x: pca.transform(x))           # Apply PCA
   
    # Con esto creamos la transformacion, pero no la aplicamos
    #pipe2.fit(_df)
    _df = pipe2.apply(_df)
    # Guardamos la tubería para poder reimplmentarla a futuro
    if (save_proc==True):
        # Lo guardamos con pickle
        pickle.dump(pipe2, open(config.FOLDER_OUT_CLUST_MODEL + "/transformation_model.pkl", 'wb'))
    # Finalmente se aplica la transformacon sobre los datos de entrenamiento
    #_df = pipe2.apply(_df)

    _df=pd.DataFrame(_df)
    print('Final shape:\t', _df.shape)
    return _df
```
	
- The steps for the pipeline are clear but I have doubts about the fit and apply stages of the pipeline. For one, this code only works if I fit the pipe1 block and apply the pipe2 block.

- If I try to fit the pipe 2 before running apply the following error (occurs when calling the apply function) appears:

**PipelineApplicationError: Exception raised in stage [ 0] PdPipelineStage: One-hot encode First_OperatingUnit, Sector, Industria**

**TypeError: unhashable type: 'Series'**

- This error does not appear if I only fit the pipe 1 and apply pipe 2 directly without fitting it first but comes back in this other function that implements the transformations described in the pipeline by opening a pickle file containing the pipeline.

```
def implement_transformation(_df,_model_name):
   
    # Seleccionamos únicamente las columnas de este escenario
    cols = translate_scenario_cluster(scenario=config.VAR_WEIGHT, scenarios_file=config.ELEM_FILENAME)
    _df = _df[cols['model']].copy()
    print(_df.columns)

    # Agrupación de variables por tipo de dato
    #_df, (cols_num, cols_cat, cols_oth) = P.fix_columns_dtypes(_df)
    print(_df.columns)
    _df, cols_fact, cols_ohe, dict_weights, cols_num, cols_model = fix_df(_df)

   
    # Se carga el modelo de cluster
    pipeline = pickle.load(open(config.FOLDER_OUT_CLUST_MODEL+'/'+ _model_name, 'rb'))
    # Lo guardamos para trazabilidad
    pickle.dump(pipeline, open(config.FOLDER_OUT_EXEC_TRANS + "/" + _model_name, 'wb'))
   
    # Se aplica la transformacion que se ha leido del fichero de configuracion
    transformed_data = pipeline.apply(_df)
    # Transformamos a DF
    transformed_data=pd.DataFrame(transformed_data)
    return transformed_data

```
- The error comes back (exact same error) when calling the apply in this function. I have looked for similar errors but haven't found any answer to this problem so I thought maybe someone here could help me.
****

**Comment by shaypal5** [comment, marked as answer]
First of all, regarding the `pipe1.fit(_df)` line: Please note that pipelines never change dataframes in place, so to have the pipeline transform the dataframe, you'll have to use the form:

```python
_df = pipe1.fit_transform(_df)
```

Alternatively, you can use the ``apply`` logic using either `_df = pipe1.apply(_df)` or `_df = pipe1(_df)`. When called for an unfitted pipeline, `apply` calls `fit_transform`; when the pipeline is fitted, it calls transform.

**Reply by germanlinarespardo** [reply]
Thanks for the answer.

My issue with that line is that since I have set up 2 different pipelines to accomodate for the PCA stage (it has a fit and a transform step), which in itself I don't know if it is the correct decision, I figured I just needed to fit it in the first pipeline and apply it in the second. Thanks for the correction.

However, the error I posted does not appear when calling the first function. It comes up whenever I call the second function which opens the pickle with the information saved from the first function. I want to use the information in the pickle to apply it to new sets of information but in the `transformed_data = pipeline.apply(_df)` from the second function the error comes up and I have no idea how to fix it.

**Comment by shaypal5** [comment]
I'm in the middle of adding some functionality to help with your usecase.

You should be able to create a single pipeline to fit and transform.

Give me a few more hours... :)

**Reply by germanlinarespardo** [reply]
That is great to hear and although our problem with PCA was solved utilizing this 2 pipeline system it will be nice to have the option to do it at once.

Sorry for insisting, but our main problem is that once we have fitted the pipe, whenever we want to do an apply (underneath it is calling the transform function), we are getting the error posted above, which has something to do with the OHE step.

As an example, if we call fit, then transform, we get the error when calling transform. However, if we call apply directly without having fitted the pipe before we don't get the error until the next transform/apply. I guess it has something to do with how the OHE function handles the data when repeatedly applied.

**Comment by shaypal5** [comment]
Oh, I think I get what you're trying to do.

You don't need to create two pipelines, just a single one, which you can fit and transform in one call.

I've added the `Decompose` pipeline stage to each your use of PCA, but you could have done this by extending the `PdPipelineStage` class (your use case does not fit an `AdHocStage`, as you need to save the `PCA` object as an attribute).

```python
pipeline = pdp.PdPipeline([]) # PipeLine de entrenamiento
    if(config.use_OHE):          pipe1 += pdp.OneHotEncode(columns=cols_ohe, exclude_columns=None)    # One Hot Encoding
    if(config.use_FACT):         pipe1 += pdp.Encode(columns=cols_fact, exclude_columns=None)         # Factorize Encoding
    if(config.use_Scaler):       pipe1 += pdp.Scale(config.SCALER)    # Standard Scaler    , exclude_columns = cols_cat + cols_oth)    # Standard Scaler          
    if(config.use_weights):      pipe1 += pdp.AdHocStage(lambda x: apply_weights(x, dict_weights, cols_model, cols_ohe)) # Apply custom weights
    if(config.use_PCA):          pipe1 += pdp.Decompose(PCA(), n_components=config.PCA_nvars))           
```

Then just call `transformed_data = pipeline.fit_transform(_df)`.

I wouldn't use `apply` in your case. It's a nice abstraction, but in your case you know in each line whether you want to `fit_transform` or to `transform`, so be explicit about it!

**Comment by shaypal5** [comment]
The `Decompose()` stage is released in version `v0.2.7`:
https://github.com/pdpipe/pdpipe/releases/tag/v0.2.7

**Comment by shaypal5** [comment]
So, two things:

1. Can I get the full stack trace of the error?

2. Wanna jump on a quick Zoom/Meet call? I think we an figure this out in like 5 minutes if we don't do this asynchronously.

**Reply by germanlinarespardo** [reply]
1. Of course. Here it is:
2. Sure. That would be great. If you are not able to do it today we can agree on a time for tomorrow that is good for you.

Edit: Sorry I had posted the wrong error, it is updated now.
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File ~\Anaconda3\envs\WOP\lib\site-packages\pdpipe\core.py:1176, in PdPipeline.transform(self, X, y, exraise, verbose, time, application_context)
   1175     stage.application_context = self.application_context
-> 1176     inter_df = stage.transform(
   1177         X=inter_df,
   1178         y=None,
   1179         exraise=exraise,
   1180         verbose=verbose,
   1181     )
   1182 except Exception as e:

File ~\Anaconda3\envs\WOP\lib\site-packages\pdpipe\core.py:490, in PdPipelineStage.transform(self, X, y, exraise, verbose)
    488     raise UnfittedPipelineStageError(
    489         "transform of an unfitted pipeline stage was called!")
--> 490 res_df = self._transform(X, verbose=verbose)
    491 if exraise and not self._compound_post(df=res_df):

File ~\Anaconda3\envs\WOP\lib\site-packages\pdpipe\col_generation.py:285, in OneHotEncode._transform(self, df, verbose)
    281     raise PipelineApplicationError((
    282         "Missing encoder for column {} when applying a fitted "
    283         "OneHotEncode pipeline stage by class {} !")
    284         .format(colname, self.__class__))
--> 285 res_cols = col.apply(encoder)
    286 for res_col in res_cols:

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\core\series.py:4433, in Series.apply(self, func, convert_dtype, args, **kwargs)
   4330 """
   4331 Invoke function on values of Series.
   4332 
   (...)
   4431 dtype: float64
   4432 """
-> 4433 return SeriesApply(self, func, convert_dtype, args, kwargs).apply()

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\core\apply.py:1082, in SeriesApply.apply(self)
   1080     return self.apply_str()
-> 1082 return self.apply_standard()

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\core\apply.py:1129, in SeriesApply.apply_standard(self)
   1127 if is_extension_array_dtype(obj.dtype) and hasattr(obj._values, "map"):
   1128     # GH#23179 some EAs do not have `map`
-> 1129     mapped = obj._values.map(f)
   1130 else:

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\core\arrays\categorical.py:1432, in Categorical.map(self, mapper)
   1431 try:
-> 1432     return self.from_codes(
   1433         self._codes.copy(), categories=new_categories, ordered=self.ordered
   1434     )
   1435 except ValueError:
   1436     # NA values are represented in self._codes with -1
   1437     # np.take causes NA values to take final element in new_categories

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\core\arrays\categorical.py:688, in Categorical.from_codes(cls, codes, categories, ordered, dtype)
    650 """
    651 Make a Categorical type from codes and categories or dtype.
    652 
   (...)
    686 Categories (2, object): ['a' < 'b']
    687 """
--> 688 dtype = CategoricalDtype._from_values_or_dtype(
    689     categories=categories, ordered=ordered, dtype=dtype
    690 )
    691 if dtype.categories is None:

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\core\dtypes\dtypes.py:296, in CategoricalDtype._from_values_or_dtype(cls, values, categories, ordered, dtype)
    292 else:
    293     # If dtype=None and values is not categorical, create a new dtype.
    294     # Note: This could potentially have categories=None and
    295     # ordered=None.
--> 296     dtype = CategoricalDtype(categories, ordered)
    298 return cast(CategoricalDtype, dtype)

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\core\dtypes\dtypes.py:183, in CategoricalDtype.__init__(self, categories, ordered)
    182 def __init__(self, categories=None, ordered: Ordered = False):
--> 183     self._finalize(categories, ordered, fastpath=False)

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\core\dtypes\dtypes.py:337, in CategoricalDtype._finalize(self, categories, ordered, fastpath)
    336 if categories is not None:
--> 337     categories = self.validate_categories(categories, fastpath=fastpath)
    339 self._categories = categories

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\core\dtypes\dtypes.py:533, in CategoricalDtype.validate_categories(categories, fastpath)
    531     raise ValueError("Categorical categories cannot be null")
--> 533 if not categories.is_unique:
    534     raise ValueError("Categorical categories must be unique")

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\_libs\properties.pyx:37, in pandas._libs.properties.CachedProperty.__get__()

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\core\indexes\base.py:2237, in Index.is_unique(self)
   2234 """
   2235 Return if the index has unique values.
   2236 """
-> 2237 return self._engine.is_unique

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\_libs\index.pyx:223, in pandas._libs.index.IndexEngine.is_unique.__get__()

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\_libs\index.pyx:230, in pandas._libs.index.IndexEngine._do_unique_check()

File ~\Anaconda3\envs\WOP\lib\site-packages\pandas\_libs\index.pyx:287, in pandas._libs.index.IndexEngine._ensure_mapping_populated()

File pandas\_libs\hashtable_class_helper.pxi:5231, in pandas._libs.hashtable.PyObjectHashTable.map_locations()

TypeError: unhashable type: 'Series'

The above exception was the direct cause of the following exception:

PipelineApplicationError                  Traceback (most recent call last)
Input In [490], in <cell line: 4>()
      2 df_test = df.copy()
      3 # APlicamos las transformaciones para corroboar que son iguales que las del entrenamiento
----> 4 df_test_trans = implement_transformation(df_test, "transformation_model.pkl")
      5 df_test_trans.head(10)

Input In [489], in implement_transformation(_df, _model_name)
     14 pickle.dump(pipeline, open(config.FOLDER_OUT_EXEC_TRANS + "/" + _model_name, 'wb'))
     16 # Se aplica la transformacion que se ha leido del fichero de configuracion
---> 17 transformed_data = pipeline(_df)
     18 transformed_data = pd.DataFrame(transformed_data, index=df.index)
     20 # Transformamos a DF

File ~\Anaconda3\envs\WOP\lib\site-packages\pdpipe\core.py:890, in PdPipeline.apply(self, df, exraise, verbose, time, fit_context, application_context)
    852 """Applies this pipeline stage to the given dataframe.
    853 
    854 If the stage is not fitted fit_transform is called. Otherwise,
   (...)
    887     The resulting dataframe.
    888 """
    889 if self.is_fitted:
--> 890     res = self.transform(
    891         X=df,
    892         exraise=exraise,
    893         verbose=verbose,
    894         time=time,
    895         application_context=application_context,
    896     )
    897     return res
    898 res = self.fit_transform(
    899     X=df,
    900     exraise=exraise,
   (...)
    904     application_context=application_context,
    905 )

File ~\Anaconda3\envs\WOP\lib\site-packages\pdpipe\core.py:1183, in PdPipeline.transform(self, X, y, exraise, verbose, time, application_context)
   1176         inter_df = stage.transform(
   1177             X=inter_df,
   1178             y=None,
   1179             exraise=exraise,
   1180             verbose=verbose,
   1181         )
   1182     except Exception as e:
-> 1183         raise PipelineApplicationError(
   1184             f"Exception raised in stage [ {i}] {stage}"
   1185         ) from e
   1186 self._post_transform_lock()
   1187 return inter_df

PipelineApplicationError: Exception raised in stage [ 0] PdPipelineStage: One-hot encode Pais, Sector

---------------------------------------------------------------------------

**Comment by shaypal5** [comment]
Hey!

When is a good time for you during the next six hours?
@germanlinarespardo

**Comment by shaypal5** [comment]
Hey @germanlinarespardo 

The OneHotEncode stage now works for Categorical dtype columns.
Released in `v0.2.8`.

Please check and let me know how it works! :)

**Reply by germanlinarespardo** [reply]
Hey Shay,

It works perfectly fine. Thanks a lot for all the effort!

Cheers

**Reply by shaypal5** [reply]
Yay! Good to hear. :)
